#include "common_dbc.h"

namespace {

const Signal sigs_33[] = {
    {
      .name = "CHECKSUM",
      .b1 = 56,
      .b2 = 8,
      .bo = 0,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "BRAKE_ENGAGED",
      .b1 = 2,
      .b2 = 1,
      .bo = 61,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "BRAKE_PRESSURE",
      .b1 = 48,
      .b2 = 8,
      .bo = 8,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_37[] = {
    {
      .name = "STEER_ANGLE",
      .b1 = 41,
      .b2 = 23,
      .bo = 0,
      .is_signed = true,
      .factor = 0.00039246467,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_71[] = {
    {
      .name = "CHECKSUM",
      .b1 = 56,
      .b2 = 8,
      .bo = 0,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "APPS_1",
      .b1 = 11,
      .b2 = 13,
      .bo = 40,
      .is_signed = false,
      .factor = 0.00015625,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "APPS_2",
      .b1 = 27,
      .b2 = 13,
      .bo = 24,
      .is_signed = false,
      .factor = 0.00033955857,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_76[] = {
    {
      .name = "GAS_ENGAGED",
      .b1 = 6,
      .b2 = 1,
      .bo = 57,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_85[] = {
    {
      .name = "CHECKSUM",
      .b1 = 48,
      .b2 = 8,
      .bo = 8,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "WHEELSPEED_F",
      .b1 = 0,
      .b2 = 24,
      .bo = 40,
      .is_signed = false,
      .factor = 1.37e-05,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "WHEELSPEED_B",
      .b1 = 24,
      .b2 = 24,
      .bo = 16,
      .is_signed = false,
      .factor = 3.6e-06,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_88[] = {
    {
      .name = "CHECKSUM",
      .b1 = 32,
      .b2 = 8,
      .bo = 24,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "GEAR",
      .b1 = 7,
      .b2 = 6,
      .bo = 51,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_107[] = {
    {
      .name = "MAIN_TORQUE",
      .b1 = 8,
      .b2 = 8,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SUB_TORQUE",
      .b1 = 24,
      .b2 = 8,
      .bo = 32,
      .is_signed = false,
      .factor = 1,
      .offset = -111.0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_384[] = {
    {
      .name = "CHECKSUM",
      .b1 = 24,
      .b2 = 8,
      .bo = 32,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "HANDBRAKE_ENGAGED",
      .b1 = 0,
      .b2 = 1,
      .bo = 63,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_512[] = {
    {
      .name = "GAS_COMMAND",
      .b1 = 0,
      .b2 = 16,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "GAS_COMMAND2",
      .b1 = 16,
      .b2 = 16,
      .bo = 32,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "ENABLE",
      .b1 = 47,
      .b2 = 1,
      .bo = 16,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "COUNTER_PEDAL",
      .b1 = 52,
      .b2 = 4,
      .bo = 8,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::PEDAL_COUNTER,
    },
    {
      .name = "CHECKSUM_PEDAL",
      .b1 = 56,
      .b2 = 8,
      .bo = 0,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::PEDAL_CHECKSUM,
    },
};
const Signal sigs_513[] = {
    {
      .name = "INTERCEPTOR_GAS",
      .b1 = 0,
      .b2 = 16,
      .bo = 48,
      .is_signed = false,
      .factor = 0.0007,
      .offset = -0.342,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "INTERCEPTOR_GAS2",
      .b1 = 16,
      .b2 = 16,
      .bo = 32,
      .is_signed = false,
      .factor = 0.0007,
      .offset = -0.691,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "STATE",
      .b1 = 36,
      .b2 = 4,
      .bo = 24,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_514[] = {
    {
      .name = "INTERCEPTOR_MAIN_TORQUE",
      .b1 = 0,
      .b2 = 16,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "INTERCEPTOR_SUB_TORQUE",
      .b1 = 16,
      .b2 = 16,
      .bo = 32,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "DIRECTION",
      .b1 = 39,
      .b2 = 1,
      .bo = 24,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "ENABLE",
      .b1 = 47,
      .b2 = 1,
      .bo = 16,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "COUNTER_STEERING",
      .b1 = 52,
      .b2 = 4,
      .bo = 8,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "CHECKSUM_STEERING",
      .b1 = 56,
      .b2 = 8,
      .bo = 0,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_679[] = {
    {
      .name = "SET_ME_X40",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_X1A_1",
      .b1 = 8,
      .b2 = 8,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_XFF_1",
      .b1 = 16,
      .b2 = 8,
      .bo = 40,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_XFF_2",
      .b1 = 24,
      .b2 = 8,
      .bo = 32,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_XFF_3",
      .b1 = 32,
      .b2 = 8,
      .bo = 24,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_XFF_4",
      .b1 = 40,
      .b2 = 8,
      .bo = 16,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_X1A_2",
      .b1 = 48,
      .b2 = 8,
      .bo = 8,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_X21",
      .b1 = 56,
      .b2 = 8,
      .bo = 0,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_680[] = {
    {
      .name = "CHECKSUM",
      .b1 = 56,
      .b2 = 8,
      .bo = 0,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "AEB_PUMP_HOLD",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "MAGNITUDE",
      .b1 = 8,
      .b2 = 8,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "BRAKE_REQ",
      .b1 = 16,
      .b2 = 8,
      .bo = 40,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_XE5",
      .b1 = 24,
      .b2 = 8,
      .bo = 32,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_X1B",
      .b1 = 32,
      .b2 = 8,
      .bo = 24,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_X00_1",
      .b1 = 40,
      .b2 = 8,
      .bo = 16,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_X00_2",
      .b1 = 48,
      .b2 = 8,
      .bo = 8,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_681[] = {
    {
      .name = "CHECKSUM",
      .b1 = 48,
      .b2 = 8,
      .bo = 8,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "AEB_ALARM",
      .b1 = 15,
      .b2 = 1,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_1267[] = {
    {
      .name = "SET_ME_X23",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_X01_1",
      .b1 = 16,
      .b2 = 8,
      .bo = 40,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_ME_X01_2",
      .b1 = 24,
      .b2 = 8,
      .bo = 32,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_1584[] = {
    {
      .name = "CHECKSUM",
      .b1 = 40,
      .b2 = 8,
      .bo = 16,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "LEFT_FRONT_DOOR",
      .b1 = 4,
      .b2 = 1,
      .bo = 59,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "RIGHT_BACK_DOOR",
      .b1 = 3,
      .b2 = 1,
      .bo = 60,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "LEFT_BACK_DOOR",
      .b1 = 2,
      .b2 = 1,
      .bo = 61,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "MAIN_DOOR",
      .b1 = 14,
      .b2 = 1,
      .bo = 49,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SEAT_BELT_WARNING",
      .b1 = 13,
      .b2 = 1,
      .bo = 50,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "LEFT_SIGNAL",
      .b1 = 9,
      .b2 = 1,
      .bo = 54,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "RIGHT_SIGNAL",
      .b1 = 8,
      .b2 = 1,
      .bo = 55,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "HIGH_BEAM",
      .b1 = 24,
      .b2 = 1,
      .bo = 39,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_1616[] = {
    {
      .name = "CHECKSUM",
      .b1 = 56,
      .b2 = 8,
      .bo = 0,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "GENERIC_TOGGLE",
      .b1 = 5,
      .b2 = 1,
      .bo = 58,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "FOG_LIGHT",
      .b1 = 3,
      .b2 = 1,
      .bo = 60,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "LOW_BEAM",
      .b1 = 2,
      .b2 = 1,
      .bo = 61,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "EVENING_LIGHT",
      .b1 = 1,
      .b2 = 1,
      .bo = 62,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};

const Msg msgs[] = {
  {
    .name = "BRAKE",
    .address = 0x21,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_33),
    .sigs = sigs_33,
  },
  {
    .name = "STEERING_ANGLE_SENSOR",
    .address = 0x25,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_37),
    .sigs = sigs_37,
  },
  {
    .name = "GAS_PEDAL",
    .address = 0x47,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_71),
    .sigs = sigs_71,
  },
  {
    .name = "GAS_PEDAL_1",
    .address = 0x4C,
    .size = 7,
    .num_sigs = ARRAYSIZE(sigs_76),
    .sigs = sigs_76,
  },
  {
    .name = "WHEEL_SPEED",
    .address = 0x55,
    .size = 7,
    .num_sigs = ARRAYSIZE(sigs_85),
    .sigs = sigs_85,
  },
  {
    .name = "TRANSMISSION",
    .address = 0x58,
    .size = 5,
    .num_sigs = ARRAYSIZE(sigs_88),
    .sigs = sigs_88,
  },
  {
    .name = "STEERING_TORQUE",
    .address = 0x6B,
    .size = 4,
    .num_sigs = ARRAYSIZE(sigs_107),
    .sigs = sigs_107,
  },
  {
    .name = "HANDBRAKE",
    .address = 0x180,
    .size = 5,
    .num_sigs = ARRAYSIZE(sigs_384),
    .sigs = sigs_384,
  },
  {
    .name = "GAS_COMMAND",
    .address = 0x200,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_512),
    .sigs = sigs_512,
  },
  {
    .name = "GAS_SENSOR",
    .address = 0x201,
    .size = 6,
    .num_sigs = ARRAYSIZE(sigs_513),
    .sigs = sigs_513,
  },
  {
    .name = "TORQUE_COMMAND",
    .address = 0x202,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_514),
    .sigs = sigs_514,
  },
  {
    .name = "FWD_CAM1",
    .address = 0x2A7,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_679),
    .sigs = sigs_679,
  },
  {
    .name = "ADAS_AEB",
    .address = 0x2A8,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_680),
    .sigs = sigs_680,
  },
  {
    .name = "ADAS_HUD",
    .address = 0x2A9,
    .size = 7,
    .num_sigs = ARRAYSIZE(sigs_681),
    .sigs = sigs_681,
  },
  {
    .name = "ADAS_HEARTBEAT",
    .address = 0x4F3,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_1267),
    .sigs = sigs_1267,
  },
  {
    .name = "METER_CLUSTER",
    .address = 0x630,
    .size = 6,
    .num_sigs = ARRAYSIZE(sigs_1584),
    .sigs = sigs_1584,
  },
  {
    .name = "RIGHT_STALK",
    .address = 0x650,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_1616),
    .sigs = sigs_1616,
  },
};

const Val vals[] = {
    {
      .name = "GEAR",
      .address = 0x58,
      .def_val = "32 P 16 R 8 N 4 D 2 D 1 D",
      .sigs = sigs_88,
    },
    {
      .name = "STATE",
      .address = 0x201,
      .def_val = "5 FAULT_TIMEOUT 4 FAULT_STARTUP 3 FAULT_SCE 2 FAULT_SEND 1 FAULT_BAD_CHECKSUM 0 NO_FAULT",
      .sigs = sigs_513,
    },
    {
      .name = "STATE",
      .address = 0x202,
      .def_val = "5 FAULT_TIMEOUT 4 FAULT_STARTUP 3 FAULT_SCE 2 FAULT_SEND 1 FAULT_BAD_CHECKSUM 0 NO_FAULT",
      .sigs = sigs_514,
    },
};

}

const DBC perodua_general_pt = {
  .name = "perodua_general_pt",
  .num_msgs = ARRAYSIZE(msgs),
  .msgs = msgs,
  .vals = vals,
  .num_vals = ARRAYSIZE(vals),
};

dbc_init(perodua_general_pt)